test/distributed/_composable/fsdp/test_fully_shard_extensions.py

```diff
@@ -68,6 +76,95 @@ def two_tensor_fsdp_post_all_gather(
     return two_tensor, tensors_to_free
 
 
+class BFloat16AllGatherTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, data: torch.Tensor, pad_in_pre_all_gather: bool = True):
+        return torch.Tensor._make_wrapper_subclass(
+            cls,
+            data.shape,
+            data.stride(),
+            data.storage_offset(),
+            dtype=data.dtype,
+            device=data.device,
+        )
+
+    def __init__(self, data: torch.Tensor, pad_in_pre_all_gather: bool = True):
+        self._data = data
+        self._pad_in_pre_all_gather = pad_in_pre_all_gather
+
+    def fsdp_pre_all_gather(
+        self,
+        mesh: DeviceMesh,
+        outer_size: torch.Size,
+        outer_stride: Tuple[int, ...],
+        module: nn.Module,
+        mp_policy: MixedPrecisionPolicy,
+    ) -> Tuple[Tuple[torch.Tensor, ...], Any]:
+        assert mesh.ndim == 1, f"{mesh.ndim}"
+        mesh_size = mesh.size()
+        requires_padding = outer_size[0] % mesh_size != 0
+        if requires_padding and self._pad_in_pre_all_gather:
+            sharded_padded_size = list(outer_size)
+            sharded_padded_size[0] = math.ceil(outer_size[0] / mesh_size)
+            padded_out = torch.empty(
+                sharded_padded_size, dtype=torch.bfloat16, device=self.device
+            )
+            padded_out[: self._data.size(0)].copy_(self._data)
+            return (padded_out,), None
+        else:
+            return self._data.to(torch.bfloat16), None
+
+    def fsdp_post_all_gather(
+        self,
+        all_gather_outputs: Tuple[torch.Tensor, ...],
+        metadata: Any,
+        param_dtype: torch.dtype,
+        *,
+        out: Optional[torch.Tensor] = None,
+    ) -> Union[Tuple[torch.Tensor, Tuple[torch.Tensor, ...]], None]:
+        assert metadata is None, f"{metadata}"
+        (tensor,) = all_gather_outputs
```

Comment:
curious where we removed the padding? the unit test works so it should be handled.

Reply:
Ah, good question. FSDP will trim the padding for you given the original size of the parameter.